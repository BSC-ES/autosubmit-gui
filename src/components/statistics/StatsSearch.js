import React, { useState, useContext, Fragment } from "react";
import StatsContext from "../context/statistics/statsContext";
import ExperimentContext from "../context/experiment/experimentContext";
import Chart from "react-google-charts";
import Spinner from "../layout/Spinner";
import { DEBUG } from "../context/vars";

const StatsSearch = () => {
  const statsContext = useContext(StatsContext);
  const experimentContext = useContext(ExperimentContext);

  const {
    getExperimentStats,
    statdata,
    loading,
    clearStats,
    isError,
    errorMessage,
    totaldata,
    ticksdata,
  } = statsContext;
  const { experiment } = experimentContext;

  // const data = [
  //     ["Year", "Visitations", "Crashes", { role: "style" }],
  //     ["2010", 10, 30, "color: gray"],
  //     ["2020", 14, 25, "color: #76A7FA"],
  //     ["2030", 16, 24, "color: blue"],
  //     ["2040", 22, 32, "stroke-color: #703593; stroke-width: 4; fill-color: #C5A5CF"],
  //     [
  //       "2050",
  //       28,
  //       23,
  //       "stroke-color: #871B47; stroke-opacity: 0.6; stroke-width: 8; fill-color: #BC5679; fill-opacity: 0.2"
  //     ]
  // ];

  // const jsonTicks = JSON.stringify({ticksdata});
  // console.log(jsonTicks['ticksdata'])
  const options = {
    chartArea: { width: "90%", height: "700px" },
    hAxis: {
      title: "Jobs",
      ticks: ticksdata, //ticks: [{v:1, f:'job_1-2323'}, {v:2, f:'job_other'}],
    },
    vAxis: {
      title: "Hours",
    },
    legend: {
      position: "top",
    },
    explorer: {
      actions: ["dragToZoom", "rightClickToReset"],
      axis: "horizontal",
      // maxZoomIn: 4.0,
    },
  };

  // console.log(options);

  var summary = null;
  if (totaldata && totaldata.stats && totaldata.stats.totals) {
    summary = totaldata.stats.totals.map((item, index) => (
      <li key={index}>{item}</li>
    ));
    // console.log(totaldata.stats.totals[0]);
    // for(var i = 0; i < totaldata.stats.totals.length; i++){
    //     console.log(totaldata.stats.totals[i]);
    //     summary += <li key={i}>{totaldata.stats.totals[i]}</li>
    // }
  }

  const onSubmitStats = (e) => {
    e.preventDefault();
    DEBUG && console.log(experiment.expid);
    DEBUG && console.log("Hours : " + hour);
    DEBUG && console.log("Type : " + section);
    getExperimentStats(experiment.expid, hour, section);
  };

  const onSubmitClear = (e) => {
    e.preventDefault();
    setHour("");
    setSection("");
    clearStats();
  };

  const [hour, setHour] = useState("");
  const [section, setSection] = useState("");
  const onChangeHour = (e) => setHour(e.target.value);
  const onChangeSection = (e) => setSection(e.target.value);
  const pStyle = {
    listStyleType: "none",
  };

  return (
    <Fragment>
      <div className='row'>
        <div className='col-md-4 offset-md-4 text-center'>
          {!statdata && experiment && experimentContext.loading === false && (
            <form onSubmit={onSubmitStats} className='form'>
              <div className='input-group input-group-sm'>
                <input
                  className='form-control'
                  type='text'
                  name='section'
                  placeholder='Section'
                  onChange={onChangeSection}
                />
                <input
                  className='form-control'
                  type='text'
                  name='hours'
                  placeholder='Hours'
                  onChange={onChangeHour}
                />
                <div className='input-group-append'>
                  <input
                    type='submit'
                    value='Get Statistics'
                    className='btn btn-primary'
                    disabled={loading}
                    data-toggle='tooltip' 
                    data-placement='bottom' 
                    title="Gets the statistics for the Section and Hours values provided."
                  />
                </div>
              </div>
            </form>
          )}
          {statdata && (
            <form onSubmit={onSubmitClear} className='form'>
              <div className=''>
                <input
                  type='submit'
                  value='Clear Statistics'
                  className='btn btn-dark btn-sm'
                />
              </div>
            </form>
          )}
        </div>
        <div className='col-md-12'>
          {!statdata && experiment && (
            <div className='col-md-12 p-3'>
              <p className='lead'>
                Supply a Section (Type) in the appropriate textbox to filter the
                jobs that will be included in the query. Also, you can also
                supply the Hours value that determines how many hours before the
                current time you want to query. Leave both empty and a query for
                Any Section since the date of creation of the experiment will be
                executed.
              </p>
              <p className='lead'>
                Press <span className='badge badge-primary'>Get Statistics</span>{" "}
                to generate the statistics, this will generate a Bar Chart and
                some extra statistics below. Drag the mouse inside the chart to
                zoom in; however, zoom in capabilities are not unlimited, so try
                to narrow your query.
              </p>
            </div>
          )}
          {loading && <Spinner />}
          {statdata && !isError && (
            <Chart
              chartType='ColumnChart'
              loader={<div>Loading Chart</div>}
              width={"100%"}
              height={"700px"}
              data={statdata}
              options={options}
            />
          )}
          {statdata && isError && (
            <div className='col-md-12 text-center p-3'>{errorMessage}</div>
          )}
          {totaldata && (
            <div className='col-md-8 offset-md-4 mt-2'>
              {totaldata && <ul style={pStyle}>{summary && summary}</ul>}
            </div>
          )}
        </div>
      </div>
      {/* <div className='row' style={experimentBuffer}></div> */}
    </Fragment>
  );
};

// const experimentBuffer = {
//   minHeight: "100%",
// };

export default StatsSearch;
