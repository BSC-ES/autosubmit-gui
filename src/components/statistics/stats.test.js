import React from "react";
import { unmountComponentAtNode, render } from "react-dom";
import { BrowserRouter as Router } from "react-router-dom";
import { act } from "react-dom/test-utils";
import renderer from 'react-test-renderer';
import StatsSearch from "./StatsSearch";
import ExperimentContext from "../context/experiment/experimentContext";
import StatsContext from "../context/statistics/statsContext";

let container = null;
beforeEach(() => {
  // setup DOM element as render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it("ExperimentItem renders with content", () => {
  const experiment = {expid: "t08b", branch: "master", description: "Test", owner:"gmontane", owner_id: 1226, hpc:"LOCAL", db_historic_version: 14, version:"3.11.0"};
  const result = [
    [
      "Jobs",
      "Queued",
      "Run",
      "Failed Jobs",
      "Failed Queued",
      "Fail Run"
    ],
    [
      {
        "v": 1,
        "f": "t08b_20120101_000_1_SIM"
      },
      1.8716666666666666,
      0.007222222222222222,
      0,
      0,
      0
    ],
    [
      {
        "v": 2,
        "f": "t08b_20120101_001_1_SIM"
      },
      1.871388888888889,
      0.0077777777777777776,
      0,
      0,
      0
    ],
    [
      {
        "v": 3,
        "f": "t08b_20120101_002_1_SIM"
      },
      1.8488888888888888,
      0.005,
      0,
      0,
      0
    ],
    [
      {
        "v": 4,
        "f": "t08b_20120101_003_1_SIM"
      },
      1.87,
      0.006111111111111111,
      0,
      0,
      0
    ],
    [
      {
        "v": 5,
        "f": "t08b_20120101_1_DA"
      },
      0.0011111111111111111,
      0.1297222222222222,
      0,
      0,
      0
    ],
    [
      {
        "v": 6,
        "f": "t08b_20120101_000_1_REDUCE"
      },
      0.0011111111111111111,
      0.004722222222222222,
      0,
      0,
      0
    ],
    [
      {
        "v": 7,
        "f": "t08b_20120101_001_1_REDUCE"
      },
      0.005833333333333334,
      0.0019444444444444444,
      0,
      0,
      0
    ],
    [
      {
        "v": 8,
        "f": "t08b_20120101_002_1_REDUCE"
      },
      0.008611111111111111,
      0.0033333333333333335,
      0,
      0,
      0
    ],
    [
      {
        "v": 9,
        "f": "t08b_20120101_003_1_REDUCE"
      },
      0.0125,
      0.003611111111111111,
      0,
      0,
      0
    ],
    [
      {
        "v": 10,
        "f": "t08b_20120101_1_CALC_STATS"
      },
      0.0011111111111111111,
      0.0030555555555555557,
      0,
      0,
      0
    ],
    [
      {
        "v": 11,
        "f": "t08b_20120101_000_1_ARCHIVE"
      },
      0.0002777777777777778,
      0.016944444444444446,
      0,
      0,
      0
    ],
    [
      {
        "v": 12,
        "f": "t08b_20120101_001_1_ARCHIVE"
      },
      0,
      0.016944444444444446,
      0,
      0,
      0
    ],
    [
      {
        "v": 13,
        "f": "t08b_20120101_002_1_ARCHIVE"
      },
      0,
      0.016944444444444446,
      0,
      0,
      0
    ],
    [
      {
        "v": 14,
        "f": "t08b_20120101_003_1_ARCHIVE"
      },
      0,
      0.017777777777777778,
      0,
      0,
      0
    ]
  ]
  const requestResult = {
    "error": false,
    "error_message": "",
    "jobs": [
      "t08b_20120101_000_1_SIM",
      "t08b_20120101_001_1_SIM",
      "t08b_20120101_002_1_SIM",
      "t08b_20120101_003_1_SIM",
      "t08b_20120101_1_DA",
      "t08b_20120101_000_1_REDUCE",
      "t08b_20120101_001_1_REDUCE",
      "t08b_20120101_002_1_REDUCE",
      "t08b_20120101_003_1_REDUCE",
      "t08b_20120101_1_CALC_STATS",
      "t08b_20120101_000_1_ARCHIVE",
      "t08b_20120101_001_1_ARCHIVE",
      "t08b_20120101_002_1_ARCHIVE",
      "t08b_20120101_003_1_ARCHIVE"
    ],
    "otherstats": [],
    "periodo_fin": "Tue, 03 Nov 2020 18:49:00 GMT",
    "periodo_ini": "Tue, 03 Nov 2020 16:49:00 GMT",
    "stats": {
      "fail_queued": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0
      ],
      "fail_run": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0
      ],
      "failed_jobs": [
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0
      ],
      "max_fail": 0,
      "max_time": 1.8716666666666666,
      "queued": [
        1.8716666666666666,
        1.871388888888889,
        1.8488888888888888,
        1.87,
        0.0011111111111111111,
        0.0011111111111111111,
        0.005833333333333334,
        0.008611111111111111,
        0.0125,
        0.0011111111111111111,
        0.0002777777777777778,
        0,
        0,
        0
      ],
      "run": [
        0.007222222222222222,
        0.0077777777777777776,
        0.005,
        0.006111111111111111,
        0.1297222222222222,
        0.004722222222222222,
        0.0019444444444444444,
        0.0033333333333333335,
        0.003611111111111111,
        0.0030555555555555557,
        0.016944444444444446,
        0.016944444444444446,
        0.016944444444444446,
        0.017777777777777778
      ],
      "threshold": 1,
      "totals": [
        "Period: 2020-11-03 16:49:00 ~ 2020-11-03 18:49:00",
        "Submitted (#): 14",
        "Run  (#): 14",
        "Failed  (#): 0",
        "Completed (#): 14",
        "Queueing time (h): 7.49",
        "Expected consumption real (h): 13.33",
        "Expected consumption CPU time (h): 366.33",
        "Consumption real (h): 0.24",
        "Consumption CPU time (h): 35.07",
        "Consumption (%): 9.57"
      ]
    }
  }
  const ticks = [
    {
      "v": 1,
      "f": "t08b_20120101_000_1_SIM"
    },
    {
      "v": 2,
      "f": "t08b_20120101_001_1_SIM"
    },
    {
      "v": 3,
      "f": "t08b_20120101_002_1_SIM"
    },
    {
      "v": 4,
      "f": "t08b_20120101_003_1_SIM"
    },
    {
      "v": 5,
      "f": "t08b_20120101_1_DA"
    },
    {
      "v": 6,
      "f": "t08b_20120101_000_1_REDUCE"
    },
    {
      "v": 7,
      "f": "t08b_20120101_001_1_REDUCE"
    },
    {
      "v": 8,
      "f": "t08b_20120101_002_1_REDUCE"
    },
    {
      "v": 9,
      "f": "t08b_20120101_003_1_REDUCE"
    },
    {
      "v": 10,
      "f": "t08b_20120101_1_CALC_STATS"
    },
    {
      "v": 11,
      "f": "t08b_20120101_000_1_ARCHIVE"
    },
    {
      "v": 12,
      "f": "t08b_20120101_001_1_ARCHIVE"
    },
    {
      "v": 13,
      "f": "t08b_20120101_002_1_ARCHIVE"
    },
    {
      "v": 14,
      "f": "t08b_20120101_003_1_ARCHIVE"
    }
  ]

  act(() => {
    render(<Router><ExperimentContext.Provider value={{ experiment: {experiment}, loading: false }}><StatsContext.Provider value={{ getExperimentStatus: () => null, loading: false, clearStats: () => null, isError: false, errorMessage: "", totaldata: requestResult, ticksdata: ticks }}><StatsSearch/></StatsContext.Provider></ExperimentContext.Provider></Router>, container);
  });
  expect(container.textContent).toContain("Period: 2020-11-03 16:49:00 ~ 2020-11-03 18:49:00");

  const tree = renderer.create(<Router><ExperimentContext.Provider value={{ experiment: {experiment}, loading: false }}><StatsContext.Provider value={{ getExperimentStatus: () => null, statdata: result, loading: false, clearStats: () => null, isError: false, errorMessage: "", totaldata: requestResult, ticksdata: ticks }}><StatsSearch/></StatsContext.Provider></ExperimentContext.Provider></Router>).toJSON();
  expect(tree).toMatchSnapshot();
});