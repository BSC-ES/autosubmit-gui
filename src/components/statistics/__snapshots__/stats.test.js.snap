// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperimentItem renders with content 1`] = `
<div
  className="row"
>
  <div
    className="col-md-4 offset-md-4 text-center"
  >
    <form
      className="form"
      onSubmit={[Function]}
    >
      <div
        className=""
      >
        <input
          className="btn btn-dark btn-sm"
          type="submit"
          value="Clear Statistics"
        />
      </div>
    </form>
  </div>
  <div
    className="col-md-12"
  >
    <div>
      Loading Chart
    </div>
    <div
      className="col-md-8 offset-md-4 mt-2"
    >
      <ul
        style={
          Object {
            "listStyleType": "none",
          }
        }
      >
        <li>
          Period: 2020-11-03 16:49:00 ~ 2020-11-03 18:49:00
        </li>
        <li>
          Submitted (#): 14
        </li>
        <li>
          Run  (#): 14
        </li>
        <li>
          Failed  (#): 0
        </li>
        <li>
          Completed (#): 14
        </li>
        <li>
          Queueing time (h): 7.49
        </li>
        <li>
          Expected consumption real (h): 13.33
        </li>
        <li>
          Expected consumption CPU time (h): 366.33
        </li>
        <li>
          Consumption real (h): 0.24
        </li>
        <li>
          Consumption CPU time (h): 35.07
        </li>
        <li>
          Consumption (%): 9.57
        </li>
      </ul>
    </div>
  </div>
</div>
`;
